Update to MP61 and JEE10 (#131)
OpenLiberty/guide-reactive-service-testing: README.adoc

```
endif::[]

> The example reactive application consists of the `system` and `inventory` microservices. The `system` microservice produces messages to the Kafka message broker, and the `inventory` microservice consumes messages from the Kafka message broker. You will write integration tests to see how you can use the Kafka consumer and producer client APIs to test each service. Kafka Testcontainers and JUnit have already been included as required test dependencies in your Maven `pom.xml` files for the [hotspot=dependencies file=0]`system` and [hotspot=dependencies file=1]`inventory` microservices.

Reviewer: ```suggestion
The example reactive application consists of the `system` and `inventory` microservices. The `system` microservice produces messages to the Kafka message broker, and the `inventory` microservice consumes messages from the Kafka message broker. You will write integration tests to see how you can use the Kafka consumer and producer client APIs to test each service. Kafka Testcontainers and JUnit are already included as required test dependencies in your Maven `pom.xml` files for the [hotspot=dependencies file=0]`system` and [hotspot=dependencies file=1]`inventory` microservices.
```

The `start` directory contains three directories: the `system` microservice directory, the `inventory` microservice directory, and the `models` directory. The `models` directory contains the model class that defines the structure of the system load data that is used in the application. Run the following Maven goal to install the packaged `models` artifact to the local Maven repository so it can be used later by the `system` and `inventory` microservices:

[role='command']
```
mvn -pl models install
```

> === Launching the system microservice in dev mode with container

Reviewer: ```suggestion
=== Launching the system microservice in dev mode with container support
```

> Initiate the microservices in dev mode by executing the following command to launch a Kafka instance replicating the production environment. The `startKafka` script will launch a local Kafka container and establish a [hotspot=dockerNetworkSetup file=3]`reactive-app` network that allows the `system` and `inventory` microservices to connect to the Kafka message broker.

Reviewer: ```suggestion
Start the microservices in dev mode by running the following command to launch a Kafka instance that replicates the production environment. The `startKafka` script launches a local Kafka container. It also establishes a [hotspot=dockerNetworkSetup file=3]`reactive-app` network that allows the `system` and `inventory` microservices to connect to the Kafka message broker.
```

include::{common-includes}/os-tabs.adoc[]

[.tab_content.windows_section]
--
[role='command']
```
.\scripts\startKafka.bat
```
--

[.tab_content.mac_section.linux_section]
--
[role='command']
```
./scripts/startKafka.sh
```
--

Navigate to the `start/system` directory.

ifdef::cloud-hosted[]
```bash
cd /home/project/guide-reactive-service-testing/start/system
```

> In this IBM Cloud environment, you need to pre-create the ***logs*** directory by running the following commands:

Reviewer: ```suggestion
In this IBM Cloud environment, you must first create the ***logs*** directory by running the following commands:
```

```bash
mkdir -p /home/project/guide-reactive-service-testing/start/system/target/liberty/wlp/usr/servers/defaultServer/logs
chmod 777 /home/project/guide-reactive-service-testing/start/system/target/liberty/wlp/usr/servers/defaultServer/logs
```
endif::[]

> To launch the `system` microservice in dev mode with container, configure the container by specifying the options within [hotspot=devc_config file=0]`<containerRunOpts>` configuration for connecting to the [hotspot=reactive-app file=0]`reactive-app` network and exposing the container port.

Reviewer: ```suggestion
To launch the `system` microservice in dev mode with container support, configure the container by specifying the options within the [hotspot=devc_config file=0]`<containerRunOpts>` element to connect to the [hotspot=reactive-app file=0]`reactive-app` network and expose the container port.
```

> Run the following goal to start the `system` microservice in dev mode with container:

Reviewer: ```suggestion
Run the following goal to start the `system` microservice in dev mode with container support:
```

include::{common-includes}/os-tabs.adoc[]

[.tab_content.windows_section]
--
[role='command']
```
mvn liberty:devc
```
--

[.tab_content.mac_section.linux_section]
--
[role='command']
```
export TESTCONTAINERS_RYUK_DISABLED=true
mvn liberty:devc
```

> Read the https://java.testcontainers.org/features/configuration/#disabling-ryuk[Testcontainers custom configuration^] document for disabling Ryuk.

Reviewer: ```suggestion
For more information about disabling Ryuk, see the https://java.testcontainers.org/features/configuration/#disabling-ryuk[Testcontainers custom configuration^] document.
```

--

After you see the following message, your Liberty instance is ready in dev mode:

[role="no_copy"]
----
**************************************************************
*    Liberty is running in dev mode.
*    ...    
*    Liberty container port information:
*        Internal container HTTP port [ 9083 ] is mapped to container host port [ 9083 ] <
*   ...     
----

https://openliberty.io/docs/latest/development-mode.html[Dev mode^] holds your command-line session to listen for file changes. Open another command-line session to continue, or open the project in your editor.

> The `system` microservice actively seeks a Kafka topic for message push operations. Upon the successful initialization of a Kafka service, the `system` microservice establishes connectivity to Kafka message broker by using the [hotspot=kafkaConfig file=2]`mp.messaging.connector.liberty-kafka.bootstrap.servers` property. Additionally, the running `system` container exposes its service on the port [hotspot=systemPortExpose file=4]`9083` for testing purposes in dev mode with container.

Reviewer: ```suggestion
The `system` microservice actively seeks a Kafka topic for message push operations. After the Kafka service starts, the `system` microservice connects to the Kafka message broker by using the [hotspot=kafkaConfig file=2]`mp.messaging.connector.liberty-kafka.bootstrap.servers` property. When you run your app in dev mode with container support, the running `system` container exposes its service on the [hotspot=systemPortExpose file=4]`9083` port for testing purposes.
```

> === Implementing tests for the system microservice

Reviewer: ```suggestion
=== Testing the system microservice
```

include::finish/system/src/main/java/io/openliberty/guides/system/SystemService.java[]
----

//File 3
Dockerfile
[source, Java, linenums, role='code_column hide_tags=copyright']
----
include::finish/system/Dockerfile[]
----

> Construct the [hotspot=buildSystemImage file=0]`systemImage` using the `ImageFromDockerfile` class, which allows Testcontainers to build the Docker image from a Dockerfile during the test runtime. For instance, the provided Dockerfile at the specified paths [hotspot file=3]`./Dockerfile` is used to generate the `system:1.0-SNAPSHOT` image.

Reviewer: ```suggestion
Construct the [hotspot=buildSystemImage file=0]`systemImage` by using the `ImageFromDockerfile` class, which allows Testcontainers to build the Docker image from a Dockerfile during the test run time. For instance, the provided Dockerfile at the specified [hotspot file=3]`./Dockerfile` paths is used to generate the `system:1.0-SNAPSHOT` image.
```

Use the [hotspot=kafkaContainer file=0]`kafkaContainer` class to instantiate the `kafkaContainer` test container, initiating the `confluentinc/cp-kafka:latest` Docker image. Similarly, use the `GenericContainer` class to create the [hotspot=systemContainer file=0]`systemContainer` test container, starting the `system:1.0-SNAPSHOT` Docker image.
 

> The [hotspot=withListener file=0]`withListener()` is configured to `kafka:19092`, as the containerized `system` microservice functions as an additional producer. Therefore, the Kafka container needs to set up a listener to accommodate this requirement. For further details on using an additional consumer or producer with Kafka container, see the https://java.testcontainers.org/modules/kafka/[official Kafka test container documentation^]

Reviewer: ```suggestion
The [hotspot=withListener file=0]`withListener()` is configured to `kafka:19092`, as the containerized `system` microservice functions as an additional producer. Therefore, the Kafka container needs to set up a listener to accommodate this requirement. For more information about using an additional consumer or producer with a Kafka container, see the https://java.testcontainers.org/modules/kafka/[Testcontainers Kafka documentation^]
```

> It's important to note that [hotspot=withListener file=0]`withListener()` has been configured to `kafka:19092`, as the containerized `system` microservice will function as an additional producer. Therefore, the Kafka container needs to set up a listener to accommodate this requirement. For further details on using additional consumer or producer with Kafka container, please visit https://java.testcontainers.org/modules/kafka/[official Kafka Test Container Documentation^]

Reviewer: ```suggestion
The [hotspot=withListener file=0]`withListener()` is configured to `kafka:19092`, as the containerized `system` microservice functions as an additional producer. Therefore, the Kafka container needs to set up a listener to accommodate this requirement. For further details on using an additional consumer or producer with Kafka container, see the https://java.testcontainers.org/modules/kafka/[official Kafka test container documentation^]
```

> Given that containers are isolated by default, facilitating communication between the `kafkaContainer` and the `systemContainer` requires placing them on the same [hotspot=network1 hotspot=network2 hotspot=network3 file=0]`network`. The [hotspot=dependsOn file=0]`dependsOn()` method is used to indicate that the `system` microservice container should commence only after ensuring the readiness of the kafka container. 

Reviewer: ```suggestion
Because containers are isolated by default, facilitating communication between the `kafkaContainer` and the `systemContainer` requires placing them on the same [hotspot=network1 hotspot=network2 hotspot=network3 file=0]`network`. The [hotspot=dependsOn file=0]`dependsOn()` method is used to indicate that the `system` microservice container starts only after ensuring the readiness of the Kafka container. 
```

> Prior to initiating the `systemContainer`, it is imperative to override the `mp.messaging.connector.liberty-kafka.bootstrap.servers` property with `kafka:19092` using the [hotspot=bootstrapServerSetup file=0]`withEnv()` method. This step is necessary due to the establishment of a listener in the Kafka container, configured to handle an additional producer.

Reviewer: ```suggestion
Before you start the `systemContainer`, you must override the `mp.messaging.connector.liberty-kafka.bootstrap.servers` property with `kafka:19092` by using the [hotspot=bootstrapServerSetup file=0]`withEnv()` method. This step creates a listener in the Kafka container that is  configured to handle an additional producer.
```

> The test uses the [hotspot=KafkaConsumer hotspot=KafkaConsumerUsage file=0]`KafkaConsumer` client API, configuring the consumer to use the [hotspot=BootstrapSetting1 file=0]`BOOTSTRAP_SERVERS_CONFIG` property with the Kafka broker address if a local `system` microservice container is present. In the absence of a local service container, it uses the [hotspot=BootstrapSetting2 file=0]`getBootstrapServers()` method to obtain the broker address from the Kafka Testcontainer. Subsequently, the consumer is set up to consume messages from the [hotspot=systemLoadTopic file=0]`system.load` topic within the `Kafka` container.

Reviewer: ```suggestion
The test uses the [hotspot=KafkaConsumer hotspot=KafkaConsumerUsage file=0]`KafkaConsumer` client API, configuring the consumer to use the [hotspot=BootstrapSetting1 file=0]`BOOTSTRAP_SERVERS_CONFIG` property with the Kafka broker address if a local `system` microservice container is present. In the absence of a local service container, it uses the [hotspot=BootstrapSetting2 file=0]`getBootstrapServers()` method to obtain the broker address from the Kafka test container. Then, the consumer is set up to consume messages from the [hotspot=systemLoadTopic file=0]`system.load` topic within the `Kafka` container.
```

> To consume messages from a stream, the messages need to be deserialized from bytes. Kafka has its own default deserializer, but a custom deserializer is provided for you. The deserializer is configured by the [hotspot=valueDeserializer file=0]`VALUE_DESERIALIZER_CLASS_CONFIG` property and is implemented in the [hotspot=SystemLoadDeserializer file=1]`SystemLoad` class. To learn more about Kafka APIs and their usage, please refer to the https://kafka.apache.org/documentation/#api[official Kafka Documentation^].

Reviewer: ```suggestion
To consume messages from a stream, the messages need to be deserialized from bytes. Kafka has its own default deserializer, but a custom deserializer is provided for you. The deserializer is configured by the [hotspot=valueDeserializer file=0]`VALUE_DESERIALIZER_CLASS_CONFIG` property and is implemented in the [hotspot=SystemLoadDeserializer file=1]`SystemLoad` class. To learn more about Kafka APIs and their usage, see the https://kafka.apache.org/documentation/#api[official Kafka Documentation^].
```

> The running `system` microservice container produces messages to the [hotspot=systemLoad file=2]`systemLoad` Kafka topic, as denoted by the [hotspot=Outgoing file=2]`@Outgoing` annotation. The [hotspot=testCpuStatus file=0]`testCpuStatus()` test method uses the [hotspot=poll file=0]`consumer.poll()` method from the `KafkaConsumer` client API to retrieve a record from Kafka every 3 seconds within a specified timeout limit. This record is produced by the system service. Subsequently, the method uses [hotspot=assert file=0]`Assertions` to verify that the polled record aligns with the expected record.

Reviewer: ```suggestion
The running `system` microservice container produces messages to the [hotspot=systemLoad file=2]`systemLoad` Kafka topic, as denoted by the [hotspot=Outgoing file=2]`@Outgoing` annotation. The [hotspot=testCpuStatus file=0]`testCpuStatus()` test method uses the [hotspot=poll file=0]`consumer.poll()` method from the `KafkaConsumer` client API to retrieve a record from Kafka every 3 seconds within a specified timeout limit. This record is produced by the system service. Then, the method uses [hotspot=assert file=0]`Assertions` to verify that the polled record aligns with the expected record.
```

[role='command']
```
mvn clean verify
```

You will see the following output:

[role='no_copy']
```
 Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 50.63 s - in it.io.openliberty.guides.system.SystemServiceIT

 Results:

 Tests run: 1, Failures: 0, Errors: 0, Skipped: 0

 --- failsafe:3.2.5:verify (verify) @ system ---
 ------------------------------------------------------------------------
 BUILD SUCCESS
 ------------------------------------------------------------------------
 Total time:  55.636 s
 Finished at: 2024-01-31T11:33:40-08:00
 ------------------------------------------------------------------------
```

== Testing with the Kafka producer client

The `inventory` microservice is tested in the same way as the `system` microservice. The only difference is that the `inventory` microservice consumes messages, which means that tests are written to use the Kafka producer client.

=== Launching the inventory microservice in dev mode with container

Navigate to the `start/inventory` directory.

ifdef::cloud-hosted[]
```bash
cd /home/project/guide-reactive-service-testing/start/inventory
```

> Pre-create the ***logs*** directory by running the following commands:

Reviewer: ```suggestion
First, create the ***logs*** directory by running the following commands:
```

```bash
mkdir -p /home/project/guide-reactive-service-testing/start/inventory/target/liberty/wlp/usr/servers/defaultServer/logs
chmod 777 /home/project/guide-reactive-service-testing/start/inventory/target/liberty/wlp/usr/servers/defaultServer/logs
```
endif::[]

> Run the following goal to start the `inventory` microservice in dev mode with container:

Reviewer: ```suggestion
Run the following goal to start the `inventory` microservice in dev mode with container support:
```

include::{common-includes}/os-tabs.adoc[]

[.tab_content.windows_section]
--
[role='command']
```
mvn liberty:devc
```
--

[.tab_content.mac_section.linux_section]
--
[role='command']
```
mvn liberty:devc
```
--

> === Building test REST client

Reviewer: ```suggestion
=== Building a test REST client
```

Create a REST client interface to access the `inventory` microservice.

ifdef::cloud-hosted[]
Open another command-line session by selecting **Terminal** > **New Terminal** from the menu of the IDE.
endif::[]

[role="code_command hotspot file=0", subs="quotes"]
----
#Create the `InventoryResourceClient` class.#
`inventory/src/test/java/it/io/openliberty/guides/inventory/InventoryResourceClient.java`
----

//File 0
InventoryResourceClient.java
[source, Java, linenums, role='code_column hide_tags=copyright']
----
include::finish/inventory/src/test/java/it/io/openliberty/guides/inventory/InventoryResourceClient.java[]
----

The `InventoryResourceClient` interface declares the [hotspot=getSystems file=0]`getSystems()` and [hotspot=resetSystems file=0]`resetSystems()` methods for accessing the corresponding endpoints within the `inventory` microservice.

> === Implementing tests for the inventory microservice

Reviewer: ```suggestion
=== Testing the inventory microservice
```

include::finish/inventory/src/main/java/io/openliberty/guides/inventory/InventoryResource.java[]
----

The [hotspot=InventoryServiceIT file=0]`InventoryServiceIT` class uses the [hotspot=KafkaProducer hotspot=KafkaProducer2 file=0]`KafkaProducer` client API to generate messages in the test environment, which are then consumed by the `inventory` microservice container.

> Similar to `system` microservice testing, the configuration of the producer [hotspot=BootstrapServerConfig hotspot=BootstrapServerConfig2 file=0]`BOOTSTRAP_SERVERS_CONFIG` property depends on whether a local `inventory` microservice container is detected.  In addition, the producer is configured with a custom serializer provided in the [hotspot=SystemLoadSerializer file=1]`SystemLoad` class.

Reviewer: ```suggestion
Similar to `system` microservice testing, the configuration of the producer [hotspot=BootstrapServerConfig hotspot=BootstrapServerConfig2 file=0]`BOOTSTRAP_SERVERS_CONFIG` property depends on whether a local `inventory` microservice container is detected. In addition, the producer is configured with a custom serializer provided in the [hotspot=SystemLoadSerializer file=1]`SystemLoad` class.
```

> The [hotspot=testCpuUsage file=0]`testCpuUsage` test method uses the [hotspot=systemLoadMsg file=0]`producer.send()` method, using the `KafkaProducer` client API, to generate the `Systemload` message. Subsequently, it uses [hotspot=assert hotspot=assert2 hotspot=assert3 file=0]`Assertions` to verify that the response from the `inventory` microservice aligns with the expected outcome.

Reviewer: ```suggestion
The [hotspot=testCpuUsage file=0]`testCpuUsage` test method uses the [hotspot=systemLoadMsg file=0]`producer.send()` method, using the `KafkaProducer` client API, to generate the `Systemload` message. Then, it uses [hotspot=assert hotspot=assert2 hotspot=assert3 file=0]`Assertions` to verify that the response from the `inventory` microservice aligns with the expected outcome.
```

[role='command']
```
mvn clean verify
```

You will see the following output:

[role='no_copy']
```
 Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 53.22 s - in it.io.openliberty.guides.inventory.InventoryServiceIT

 Results:

 Tests run: 1, Failures: 0, Errors: 0, Skipped: 0

 --- failsafe:3.2.5:verify (verify) @ inventory ---
 ------------------------------------------------------------------------
 BUILD SUCCESS
 ------------------------------------------------------------------------
 Total time:  58.789 s
 Finished at: 2024-01-31T11:40:43-08:00
 ------------------------------------------------------------------------
```

> When you're finished trying out the microservice, you can stop the local Kafka container by running the following command in `start` directory:

Reviewer: ```suggestion
When you're finished trying out the microservice, you can stop the local Kafka container by running the following command from the `start` directory:
```
